File: clojure/algos/src/bsf.clj
(ns bsf)

(defn queue
  ([] (clojure.lang.PersistentQueue/EMPTY))
  ([coll]
    (reduce conj clojure.lang.PersistentQueue/EMPTY coll)))

(def graph1
  {"you"    '("alice" "bob" "claire")
   "bob"    '("anuj" "peggy")
   "alice"  '("peggy")
   "claire" '("thom" "johny")
   "anuj"   '()
   "peggy"  '()
   "thom"   '()
   "johny"  '()
  })

(defn is-seller [n]
  (= (subs n (dec (count n))) "m"))

(defn search [graph name]
  (letfn
    [(searchInQueue [searched queue]
      (let
        [person (peek queue)
         persons (pop queue)]
        (cond
          (contains? searched person) (recur searched persons)
          (is-seller person) person
          :else (recur (conj searched person) (into persons (graph person)))
        )))]
  (searchInQueue #{} (into (queue) (graph name)))))

(defn -main []
  (println (search graph1 "you"))
  (println "[done]"))

File: haskell/algos/src/BSF.hs
module BSF where

import           Universum

import           Data.Sequence (Seq (..), (><))
import qualified Data.Sequence as Seq
import           Data.Map (Map)
import qualified Data.Map as Map
import           Data.Set (Set)
import qualified Data.Set as Set
import qualified Data.Text as Text

type Graph = Map Text [Text]

graph1 :: Graph
graph1 = Map.fromList
  [ ("you",    ["alice", "bob", "claire"])
  , ("bob",    ["anuj", "peggy"])
  , ("alice",  ["peggy"])
  , ("claire", ["thom", "johny"])
  , ("anuj",   [])
  , ("peggy",  [])
  , ("thom",   [])
  , ("johny",  [])
  ]

search :: Graph -> Text -> Maybe Text
search graph name = searchInQueue Set.empty $ personConnections name
  where
    personConnections :: Text -> Seq Text
    personConnections name = Seq.fromList . fromMaybe [] $ Map.lookup name graph

    searchInQueue :: Set Text -> Seq Text -> Maybe Text
    searchInQueue _ Empty = Nothing
    searchInQueue searched (person :<| persons)
      | Set.member person searched = searchInQueue searched persons
      | isSeller person = Just person
      | otherwise = searchInQueue (Set.insert person searched) $ persons >< personConnections person

isSeller :: Text -> Bool
isSeller name = Text.last name == 'm'

bsfTest :: IO ()
bsfTest = do
  putTextLn "BSF test"
  putTextLn $ show graph1
  putTextLn $ fromMaybe "not found" $ search graph1 "you"
  putTextLn $ fromMaybe "not found" $ search graph1 "bob"
